File: coding/Leetcode/Tasks/task_1474.cpp
// https://leetcode.com/problems/delete-n-nodes-after-m-nodes-of-a-linked-list/?envType=study-plan-v2&envId=premium-algo-100
// Created by omega515 on 12.07.24.
//

#include <bits/stdc++.h>

using namespace std;


struct ListNode {
    int val;
    ListNode *next;

    ListNode() : val(0), next(nullptr) {}

    ListNode(int x) : val(x), next(nullptr) {}

    ListNode(int x, ListNode *next) : val(x), next(next) {}
};


class Solution {
public:
    ListNode *deleteNodes(ListNode *head, int m, int n) {
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);
        cout.tie(NULL);

        ListNode *answer = head;

        ListNode *curr;
        while (head) {
            int i = 0;
            while (head && i < m - 1) {
                head = head->next;
                ++i;
            }
            if (head) {
                curr = head->next;
                i = 0;
                while (curr && i < n) {
                    head->next = curr->next;
                    curr = head->next;
                    ++i;
                }
                head = curr;
                cout << "n";
            }
        }

        return answer;
    }

};

void read_data(vector<int> h, ListNode *&head, ListNode *&curr) {
    int n = h.size();
    for (int i = 0; i < n; ++i) {
        ListNode *p = new ListNode(h[i]);
        if (curr == nullptr) {
            head = p;
            curr = p;
        } else {
            curr->next = p;
            curr = p;
        }
    }

    return;
}

vector<int> toVector(ListNode *head) {
    vector<int> v;
    while (head) {
        v.push_back(head->val);
        head = head->next;
    }
    return v;
}

int main() {
    int n, m;
    ListNode *head(nullptr), *curr(nullptr);
    vector<int> inputV;
    vector<int> answer, res;
    Solution solve;


    head = nullptr;
    curr = nullptr;
    inputV = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13}, m = 2, n = 3;
    answer = {1, 2, 6, 7, 11, 12};
    read_data(inputV, head, curr);
    solve.deleteNodes(head, m, n);
    res = toVector(head);
    assert(answer == res);


    head = nullptr;
    head = nullptr;
    curr = nullptr;
    inputV = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}, m = 1, n = 3;
    answer = {1, 5, 9};
    read_data(inputV, head, curr);
    solve.deleteNodes(head, m, n);
    res = toVector(head);
    assert(answer == res);

    return 0;
}

File: coding/Leetcode/Tasks/task_1474.java
// https://leetcode.com/problems/delete-n-nodes-after-m-nodes-of-a-linked-list/description/?envType=weekly-question&envId=2025-05-29
package leetcode.tasks;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.Queue;

public class task_1474 {
    /**
     * Definition for singly-linked list.
     * public class ListNode {
     * int val;
     * ListNode next;
     * ListNode() {}
     * ListNode(int val) { this.val = val; }
     * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
     * }
     */
    public class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    class Solution {
        public ListNode deleteNodes(ListNode head, int m, int n) {
            ListNode answer = head;
            ListNode current = head;
            ListNode prev;
            while (current != null) {
                prev = current;
                int i = 0;
                while (prev != null && i < (m - 1)) {
                    prev = prev.next;
                    ++i;
                }
                if (prev == null) {
                    break;
                }
                i = 0;
                current = prev.next;
                while (current != null && i < n) {
                    current = current.next;
                    ++i;
                }
                prev.next = current;
            }

            return answer;
        }
    }

    public ListNode read(int size) {
        ListNode prev = new ListNode(1);
        ListNode head = prev;
        for (int i = 1; i < size; i++) {
            ListNode curr = new ListNode(i + 1);
            prev.next = curr;
            prev = curr;
        }
        return head;
    }

    public void main(String[] args) {
        ListNode head;
        int m, n;
        ListNode answer;

        Solution solution = new Solution();

        int size = 11;
        head = read(size);
        m = 11;
        n = 3;
        answer = solution.deleteNodes(head, m, n);

    }
}
